### ISSUE TITLE
false positive: cppcoreguidelines-virtual-class-destructor for templated base and derived class if derived class has another virtual function

### ISSUE THREAD

**Issue by Kwasniok** [llvm/llvm-project#59032]
Heads up: This bug seems to be very specific.

## Version (14.0.6)
```console
  LLVM version 14.0.6
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: haswell
```
Unfortunately, I could not get my hands on a newer version.

## Related
- #51759
- #51254

## How to Reproduce
```cpp
// test.cpp
template <class>
struct A {
    virtual ~A() = default;
};

template <class T>
struct B : public A<T> {
    ~B() override = default;
    virtual void f() = 0;
};
```
running:
```bash
clang-tidy --checks="cppcoreguidelines-virtual-class-destructor" test.cpp -- -std=c++20
```
results in:
```console
1 warning generated.
test.cpp:8:8: warning: destructor of 'B' is public and non-virtual [cppcoreguidelines-virtual-class-destructor]
struct B : public A<T> {
       ^
test.cpp:8:8: note: make it public and virtual
struct B : public A<T> {
       ^
```

## Why This Bug is So Specific
The false positive vanishes in either of the following cases:
- base class `A` is not a template
- derived class `B` is not a template
- derived class `B` does not have another (`public` or `protected`) virtual member function
- adding the keyword `virtual` to `~B`, but this triggers `cppcoreguidelines-explicit-virtual-functions` instead 

## Further Observation
`clang++ -Werror -Wnon-virtual-dtor -c test.cpp` compiles without errors (and so does `g++`).

**Comment by EugeneZelenko** [comment]
Could you please try 15 or `main` branch?

**Comment by Kwasniok** [comment]
> Could you please try 15 or `main` branch?

I am on it, will have to install it first. So give me a moment please.

**Comment by Kwasniok** [comment]
I was finally able to test this on `main` (c6a838e9c89aac0a2d15c4114e19a29497e4d665) and it's already fixed. Thank you to all!
